File: app-android/src/main/java/com/ustadmobile/port/android/view/DashboardEntryListRecyclerAdapter.java
package com.ustadmobile.port.android.view;

import android.app.Activity;
import android.arch.paging.PagedListAdapter;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.AppCompatImageView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.PopupMenu;
import android.widget.TextView;

import com.toughra.ustadmobile.R;
import com.ustadmobile.core.controller.DashboardEntryListPresenter;
import com.ustadmobile.core.controller.ReportChartViewComponentPresenter;
import com.ustadmobile.core.controller.ReportSalesLogComponentPresenter;
import com.ustadmobile.core.controller.ReportTopLEsComponentPresenter;
import com.ustadmobile.lib.db.entities.DashboardEntry;

import java.util.Hashtable;

import static com.ustadmobile.core.view.ReportOptionsDetailView.ARG_REPORT_OPTIONS;

public class DashboardEntryListRecyclerAdapter extends
        PagedListAdapter<DashboardEntry,
                DashboardEntryListRecyclerAdapter.DashboardEntryListViewHolder> {

    Context theContext;
    Activity theActivity;
    DashboardEntryListPresenter mPresenter;

    @NonNull
    @Override
    public DashboardEntryListViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View list = LayoutInflater.from(theContext).inflate(
                R.layout.item_dashboard_entry, parent, false);
        return new DashboardEntryListViewHolder(list);

    }

    @Override
    public void onBindViewHolder(@NonNull DashboardEntryListViewHolder holder, int position) {

        DashboardEntry dashboardEntry = getItem(position);
        long entryUid = dashboardEntry.getDashboardEntryUid();
        String existingTitle = dashboardEntry.getDashboardEntryTitle();
        String entryParams = dashboardEntry.getDashboardEntryReportParam();
        int reportType = dashboardEntry.getDashboardEntryReportType();

        AppCompatImageView dots = holder.itemView.findViewById(R.id.item_dashboard_entry_dots);
        AppCompatImageView pin = holder.itemView.findViewById(R.id.item_dashboard_entry_flag);
        TextView title = holder.itemView.findViewById(R.id.item_dashboard_entry_title);

        //Title
        title.setText(existingTitle);

        //Pinned
        boolean pinned;
        if(dashboardEntry.getDashboardEntryIndex() < 0 ){
            pinned=true;
            pin.setColorFilter(ContextCompat.getColor(theContext, R.color.primary_dark));
        }else{
            pinned = false;
            pin.setColorFilter(ContextCompat.getColor(theContext, R.color.text_primary));
        }
        boolean finalPinned = pinned;
        pin.setOnClickListener(v -> mPresenter.handlePinEntry(entryUid, finalPinned));


        //Options to Edit/Delete every schedule in the list
        dots.setOnClickListener((View v) -> {
            //creating a popup menu
            PopupMenu popup = new PopupMenu(theContext, v);
            popup.setOnMenuItemClickListener(item -> {
                int i = item.getItemId();
                if (i == R.id.edit) {
                    mPresenter.handleEditEntry(entryUid);
                    return true;
                } else if (i == R.id.delete) {
                    mPresenter.handleDeleteEntry(entryUid);
                    return true;
                } else if (i == R.id.set_title) {
                    mPresenter.handleChangeTitle(entryUid, existingTitle);
                    return true;
                } else {
                    return false;
                }
            });
            //inflating menu from xml resource
            popup.inflate(R.menu.menu_edit_delete_set_title);

            popup.getMenu().findItem(R.id.edit).setVisible(true);

            //displaying the popup
            popup.show();
        });

        //The actual report
        View reportPlaceholder =
                holder.itemView.findViewById(R.id.item_dashboard_entry_report_view_placeholder);
        LinearLayout chartLL =
                holder.itemView.findViewById(R.id.item_dashboard_entry_ll);
        TextView currencyTV =
                holder.itemView.findViewById(R.id.item_dahboard_entry_currency);

        //Common Layout params for chart views to match parent.
        ViewGroup.LayoutParams params =
                new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                        ViewGroup.LayoutParams.MATCH_PARENT);
        //Creating the args to be sent to the Chart/Report view presenter
        Hashtable<String, String> args = new Hashtable<>();
        args.put(ARG_REPORT_OPTIONS, entryParams);


        //Creating a different chart/report view depending on the report type of the current
        //dashboard entry.
        switch (dashboardEntry.getDashboardEntryReportType()){
            case DashboardEntry.REPORT_TYPE_SALES_PERFORMANCE:

                //Have currency visible and placeholder gone.
                if(reportPlaceholder!=null)
                    reportPlaceholder.setVisibility(View.GONE);
                currencyTV.setVisibility(View.VISIBLE);

                //Clear before adding.
                chartLL.removeAllViews();

                //Create the chart component, sets its layout and call the presenter on the view.
                ReportSalesPerformanceChartComponent chartComponent =
                        new ReportSalesPerformanceChartComponent(theContext);
                chartComponent.setLayoutParams(params);
                ReportChartViewComponentPresenter cPresenter =
                        new ReportChartViewComponentPresenter(this,args, chartComponent);
                cPresenter.onCreate(args);

                chartLL.addView(chartComponent);
                break;
            case DashboardEntry.REPORT_TYPE_SALES_LOG:
                if(reportPlaceholder!=null)
                    reportPlaceholder.setVisibility(View.GONE);
                currencyTV.setVisibility(View.INVISIBLE);

                //Clear before adding.
                chartLL.removeAllViews();

                //Create View, presenter and add it to the view
                ReportTableListComponent salesLogComponent =
                        new ReportTableListComponent(theContext);
                salesLogComponent.setLayoutParams(params);
                ReportSalesLogComponentPresenter lPresenter =
                        new ReportSalesLogComponentPresenter(this, args, salesLogComponent);
                lPresenter.onCreate(args);

                chartLL.addView(salesLogComponent);


                break;
            case DashboardEntry.REPORT_TYPE_TOP_LES:
                if(reportPlaceholder!=null)
                    reportPlaceholder.setVisibility(View.GONE);
                currencyTV.setVisibility(View.INVISIBLE);

                //
                //Create View, presenter and add it to the view
                ReportTableListComponent topLEsComponent =
                        new ReportTableListComponent(theContext);
                topLEsComponent.setLayoutParams(params);
                ReportTopLEsComponentPresenter tPresenter =
                        new ReportTopLEsComponentPresenter(this, args, topLEsComponent);
                tPresenter.onCreate(args);

                chartLL.addView(topLEsComponent);


                break;
            default:
                chartLL.removeAllViews();
                if(reportPlaceholder!=null)
                    reportPlaceholder.setVisibility(View.VISIBLE);
                currencyTV.setVisibility(View.INVISIBLE);
                break;
        }

        chartLL.setOnClickListener(v ->
                mPresenter.handleClickReport(entryUid, entryParams, reportType));
    }


    class DashboardEntryListViewHolder extends RecyclerView.ViewHolder {
        DashboardEntryListViewHolder(View itemView) {
            super(itemView);
        }
    }

    DashboardEntryListRecyclerAdapter(
            @NonNull DiffUtil.ItemCallback<DashboardEntry> diffCallback,
            DashboardEntryListPresenter thePresenter,
            Context context) {
        super(diffCallback);
        mPresenter = thePresenter;
        theContext = context;
    }


}

File: app-android/src/main/kotlin/com/ustadmobile/port/android/view/DashboardEntryListRecyclerAdapter.kt
package com.ustadmobile.port.android.view

import android.app.Activity
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.PopupMenu
import android.widget.TextView

import androidx.appcompat.widget.AppCompatImageView
import androidx.core.content.ContextCompat
import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView

import com.toughra.ustadmobile.R
import com.ustadmobile.core.controller.DashboardEntryListPresenter
import com.ustadmobile.core.controller.ReportChartViewComponentPresenter
import com.ustadmobile.core.controller.ReportSalesLogComponentPresenter
import com.ustadmobile.core.controller.ReportTopLEsComponentPresenter
import com.ustadmobile.core.view.ReportOptionsDetailView
import com.ustadmobile.lib.db.entities.DashboardEntry

import java.util.Hashtable


class DashboardEntryListRecyclerAdapter internal constructor(
        diffCallback: DiffUtil.ItemCallback<DashboardEntry>,
        internal var mPresenter: DashboardEntryListPresenter,
        internal var theContext: Context)
    : PagedListAdapter<DashboardEntry,
        DashboardEntryListRecyclerAdapter.DashboardEntryListViewHolder>(diffCallback) {
    internal var theActivity: Activity? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DashboardEntryListViewHolder {
        val list = LayoutInflater.from(theContext).inflate(
                R.layout.item_dashboard_entry, parent, false)
        return DashboardEntryListViewHolder(list)

    }

    override fun onBindViewHolder(holder: DashboardEntryListViewHolder, position: Int) {

        val dashboardEntry = getItem(position)
        val entryUid = dashboardEntry!!.dashboardEntryUid
        val existingTitle = dashboardEntry.dashboardEntryTitle
        val entryParams = dashboardEntry.dashboardEntryReportParam
        val reportType = dashboardEntry.dashboardEntryReportType

        val dots = holder.itemView.findViewById<AppCompatImageView>(R.id.item_dashboard_entry_dots)
        val pin = holder.itemView.findViewById<AppCompatImageView>(R.id.item_dashboard_entry_flag)
        val title = holder.itemView.findViewById<TextView>(R.id.item_dashboard_entry_title)

        //Title
        title.text = existingTitle

        //Pinned
        val pinned: Boolean
        if (dashboardEntry.dashboardEntryIndex < 0) {
            pinned = true
            pin.setColorFilter(ContextCompat.getColor(theContext, R.color.primary_dark))
        } else {
            pinned = false
            pin.setColorFilter(ContextCompat.getColor(theContext, R.color.text_primary))
        }
        pin.setOnClickListener { v -> mPresenter.handlePinEntry(entryUid, pinned) }


        //Options to Edit/Delete every schedule in the list
        dots.setOnClickListener { v: View ->
            //creating a popup menu
            val popup = PopupMenu(theContext, v)
            popup.setOnMenuItemClickListener { item ->
                val i = item.itemId
                if (i == R.id.edit) {
                    mPresenter.handleEditEntry(entryUid)
                    true
                } else if (i == R.id.delete) {
                    mPresenter.handleDeleteEntry(entryUid)
                    true
                } else if (i == R.id.set_title) {
                    mPresenter.handleChangeTitle(entryUid, existingTitle!!)
                    true
                } else {
                    false
                }
            }
            //inflating menu from xml resource
            popup.inflate(R.menu.menu_edit_delete_set_title)

            popup.menu.findItem(R.id.edit).isVisible = true

            //displaying the popup
            popup.show()
        }

        //The actual report
        val reportPlaceholder = holder.itemView.findViewById<View>(R.id.item_dashboard_entry_report_view_placeholder)
        val chartLL = holder.itemView.findViewById<LinearLayout>(R.id.item_dashboard_entry_ll)
        val currencyTV = holder.itemView.findViewById<TextView>(R.id.item_dahboard_entry_currency)

        //Common Layout params for chart views to match parent.
        val params = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT)
        //Creating the args to be sent to the Chart/Report view presenter
        val args = Hashtable<String, String>()
        args[ReportOptionsDetailView.ARG_REPORT_OPTIONS] = entryParams


        //Creating a different chart/report view depending on the report type of the current
        //dashboard entry.
        when (dashboardEntry.dashboardEntryReportType) {
            DashboardEntry.REPORT_TYPE_SALES_PERFORMANCE -> {

                //Have currency visible and placeholder gone.
                if (reportPlaceholder != null)
                    reportPlaceholder.visibility = View.GONE
                currencyTV.visibility = View.VISIBLE

                //Clear before adding.
                chartLL.removeAllViews()

                //Create the chart component, sets its layout and call the presenter on the view.
                val chartComponent = ReportSalesPerformanceChartComponent(theContext)
                chartComponent.layoutParams = params
                val cPresenter = ReportChartViewComponentPresenter(this, args, chartComponent)
                cPresenter.onCreate(args)

                chartLL.addView(chartComponent)
            }
            DashboardEntry.REPORT_TYPE_SALES_LOG -> {
                if (reportPlaceholder != null)
                    reportPlaceholder.visibility = View.GONE
                currencyTV.visibility = View.INVISIBLE

                //Clear before adding.
                chartLL.removeAllViews()

                //Create View, presenter and add it to the view
                val salesLogComponent = ReportTableListComponent(theContext)
                salesLogComponent.layoutParams = params
                val lPresenter = ReportSalesLogComponentPresenter(this, args, salesLogComponent)
                lPresenter.onCreate(args)

                chartLL.addView(salesLogComponent)
            }
            DashboardEntry.REPORT_TYPE_TOP_LES -> {
                if (reportPlaceholder != null)
                    reportPlaceholder.visibility = View.GONE
                currencyTV.visibility = View.INVISIBLE

                //
                //Create View, presenter and add it to the view
                val topLEsComponent = ReportTableListComponent(theContext)
                topLEsComponent.layoutParams = params
                val tPresenter = ReportTopLEsComponentPresenter(this, args, topLEsComponent)
                tPresenter.onCreate(args)

                chartLL.addView(topLEsComponent)
            }
            else -> {
                chartLL.removeAllViews()
                if (reportPlaceholder != null)
                    reportPlaceholder.visibility = View.VISIBLE
                currencyTV.visibility = View.INVISIBLE
            }
        }

        chartLL.setOnClickListener { v -> mPresenter.handleClickReport(entryUid, entryParams!!, reportType) }
    }


    inner class DashboardEntryListViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)


}
